### ISSUE TITLE
Macbook not sleeping under Big Sur

### ISSUE THREAD

**Issue by Deftknot** [pqrs-org/Karabiner-Elements#2519]
Dear all,

I am using Karabiner since long time. As I made a fresh install for Big Sur I recognzized that after installing Karabiner it seems like the deep sleep is not working anymore. I trigger the sleep function, Macbook want's to go to sleep but something is blocking the process and it immediately wakes up again. Anybody else has the same problem?

Using a Macbook Pro (Early 2015) with Big Sur and Razer Ornata Chroma keyboard.

Thanks for your support!

**Comment by Deftknot** [comment]
Am I the only one having this issue?

**Comment by pycckuu** [comment]
I have the same thing

**Comment by Deftknot** [comment]
> I have the same thing

Sory to hear :(

I'm sad that the Dev of Karabiner is not replying to this. As stated in my last post, since I uninstalled Karabiner, deep sleep is working fine again.

**Comment by jeffoio** [comment]
same issue 
Mac mini (M1, 2020) big sur 11.1 
Karabiner-Elements 13.1.4

**Comment by shartf** [comment]
can confirm on Karabiner-Elements 13.1.5
Mac mini M1, Big Sur 11.1
If I quit Karabiner-Elements, sleep works as intended.

**Comment by bastiman85** [comment]
Similar (same?) issue.
Using an active corner putting the display to sleep, it wakes up almost exactly 10 minutes after. Every time. When I put it to sleep once more it sticks, strangely.
Quitting KE resolves the issue.

Karabiner-Elements 13.1.0
iMac 2020, Big Sur 11.1

**Comment by jeffomatic** [comment]
Same issue here: triggering sleep via a keyboard shortcut or [Alfred](https://www.alfredapp.com/) will blank the display momentarily, but the computer will wake up a second later.

- M1 Macbook Air
- Big Sur 11.2
- Karabiner 13.3.0.
- `pmset -g log`: [gist](https://gist.github.com/jeffomatic/3b2ba26900e044f3b50c354d2b695205).

Workarounds I've discovered:

- Quitting Karabiner from the Misc tab of the preferences app.
- Choosing Sleep from the Apple menu works much more effectively than me. I'm not sure how this is different to other modes of sleeping.
- Turning off the display physically.

Thanks for a great utility and let me know how else I can help!

**Comment by crawfyp1** [comment]
I have same issue with Big Sur since 11.0

**Comment by oakif** [comment]
I'm glad I found this thread. I've been having this issue for at least a month now, going all the way back to Catalina. When I put my laptop to sleep, it'll turn on again about a second later. When I press escape the screen will turn off but it'll turn back on a second later. 

Karabiner Elements 13.3.0
macOS Big Sur 11.2

**Comment by oakif** [comment]
@jeffomatic Selecting "Sleep" from the Apple menu unfortunately doesn't work for me—the same issue occurs. My external display (connected through a TB3 dock) also occasionally disconnects and reconnects after I log in, but I'm unsure if that's related to Karabiner.

**Comment by Surus-sk** [comment]
Same thing for me. I've started to notice it on version 13.1 when I moved from Intel MBP to M1 MBP. When I get it to sleep or shutdown, it starts immediately again and wakes up with error report (I was told that Karabiner causes Kernel panic). After Karabiner uninstall, problems dissappeared. Now I've tried to install back 13.3.0 version but still not fixed - kernel panic is back.

macOS version: macOS 11.2.1
Karabiner version: Karabiner-Elements 13.3.0
Your Mac hardware: MacBook Pro
Your keyboard hardware: Realforce R2 Limited

**Comment by oakif** [comment]
@Surus-sk Can you provide a bit more information about the kernel panics you are seeing? I experience them sometimes but I'm unsure if it's because of Karabiner. Since it's one of the first apps I install determining it to be the root cause of anything issue is quite difficult. I am on an Intel Mac.

**Comment by Surus-sk** [comment]
Unfortunately, I havent made a printscreen, but it looked something like this (textwise):
(https://i.stack.imgur.com/Q6k38.png)
Typical not very informative error report. It appeared after every shutdown. Mac immediately turned back on and after log in, this appeared.

I've googled a bit and found on some forums, that Karabiner might be the cause. So I've unistalled it and problem was gone (I havent tried anything else until that, so it had to be Karabiner)

**Comment by Deftknot** [comment]
I opened this bug like nearly 3 months ago and many people reported the same issue. Still no update, not even feedback from the dev that they look up this issue. Sad as this tool is very cool and helpful. 

As others already mentioned: uninstalling karabiner instantly solves the problem, therefore the problem must be within karabiner.

**Comment by Surus-sk** [comment]
True. I've just briefly searched active issues and from open ones, I've found at least two, which most likely referre to same problem. In one was offered an answer:

"In that situation, the kernel panic is most likely caused by a bug in the macOS DriverKit or HID driver. There is no workaround on Karabiner-Elements, we need to work on getting macOS fixed by reporting the issue to Apple."

So if fix is required on Apple side, there is little hope. Pity, also liked the app.

**Comment by oakif** [comment]
That's disappointing if true! I wonder if the same can be said for the sleep issue on Big Sur.

**Comment by rublev** [comment]
same macbook m1 air base model big sur 11.2

**Comment by cztomsik** [comment]
hm, I think I have this one too but weird thing is that I'm quite sure it was definitely fine month ago.
```
  cztomsik$ pmset -g assertions .
2021-03-06 00:42:57 +0100 
Assertion status system-wide:
   BackgroundTask                 0
   ApplePushServiceTask           0
   UserIsActive                   1
   PreventUserIdleDisplaySleep    0
   PreventSystemSleep             0
   ExternalMedia                  0
   InternalPreventDisplaySleep    1
   PreventUserIdleSystemSleep     1
   NetworkClientActive            0
Listed by owning process:
   pid 351(WindowServer): [0x00059a6500099b5d] 00:00:00 UserIsActive named: "com.apple.iohideventsystem.queue.tickle serviceID:100000c23 name:AppleUserHIDEventSe product:Karabiner DriverKit eventType:3"  
	Timeout will fire in 1800 secs Action=TimeoutActionRelease
   pid 295(powerd): [0x0006851b00018e1c] 24:42:39 PreventUserIdleSystemSleep named: "Powerd - Prevent sleep while display is on"  
   pid 295(powerd): [0x00059a6400099b59] 00:00:00 UserIsActive named: "com.apple.powermanagement.kernel.useractive AppleHIDTransportHIDDevice:kIOHIDSystemActivityTickle nxEvent: 0"  
	Timeout will fire in 1800 secs Action=TimeoutActionRelease
   pid 295(powerd): [0x0006858d0010823b] 00:04:12 InternalPreventDisplaySleep named: "com.apple.powermanagement.delayDisplayOff"  
	Timeout will fire in 47 secs Action=TimeoutActionTurnOff
No kernel assertions.

  cztomsik$ 
```

**Comment by mkaflowski** [comment]
I thing it only affects bluetooth devices. I have the same problem with mx keys with bluetooth connection but it works good with dongle.

**Comment by jeffomatic** [comment]
My issues appear to be fixed as of Big Sur 11.3.0 and Karabiner 13.4.0. I'm guessing root cause was the same as #2517.

**Comment by kakaum** [comment]
I'm still having this issue with my dongle, as well as bluetooth, on my M1 MBA with Big Sur 11.3 and Karabiner 13.4.0, when I use Alfred's sleep command. Selecting sleep from Apple menu doesn't show the issue.

**Comment by Deftknot** [comment]
As I opened this issue, please let me comment that I am using no bluetooth devices on my Mac, therefore a bluetooth connection cannot be the issue. At least not for my problem.

**Comment by kakaum** [comment]
Maybe I was not making myself clear with what I said above. I tried to say that I was experiencing the issue with *both* dongle and bluetooth. Sorry for any inconvenience I might have brought to you, especially @Deftknot.

**Comment by hd-deman** [comment]
same for me without bluetooth devices. 
Big Sur 11.4
KE version 13.5.0

**Comment by bergmul** [comment]
Same for me w/o bluetooth devices

Big Sur 11.4
Karabiner Elements: 13.5.0

@tekezo any idea what might be happening here?

**Comment by hd-deman** [comment]
I don't know what it was, but the problem solved after creating a new default profile

**Comment by gitityhead** [comment]
> I don't know what it was, but the problem solved after creating a new default profile

Creating a new profile does not solve the problem for me?

**Comment by rmustard** [comment]
I'm having this problem and nothing works for me.
Apple Menu > Sleep
Alfred > Sleep

$ pmset -g assertions
2021-12-03 19:28:20 -0600
Assertion status system-wide:
   BackgroundTask                 0
   ApplePushServiceTask           0
   UserIsActive                   1
   PreventUserIdleDisplaySleep    0
   PreventSystemSleep             0
   ExternalMedia                  0
   PreventUserIdleSystemSleep     1
   NetworkClientActive            0
Listed by owning process:
   pid 357(WindowServer): [0x0000511100098731] 00:00:00 UserIsActive named: "com.apple.iohideventsystem.queue.tickle serviceID:100000dfa name:AppleUserHIDEventSe product:Karabiner DriverKit eventType:3"
	Timeout will fire in 600 secs Action=TimeoutActionRelease
   pid 299(powerd): [0x00005d0a00018a20] 00:00:56 PreventUserIdleSystemSleep named: "Powerd - Prevent sleep while display is on"
   pid 299(powerd): [0x0000511100098730] 00:00:00 UserIsActive named: "com.apple.powermanagement.kernel.useractive AppleUserUSBHostHIDDevice:kIOHIDSystemActivityTickle nxEvent: 0x"
	Timeout will fire in 600 secs Action=TimeoutActionRelease
   pid 347(bluetoothd): [0x00005d0a00098a1c] 00:00:56 UserIsActive named: "Bluetooth LE HID Activity"
	Timeout will fire in 543 secs Action=TimeoutActionRelease
Kernel Assertions: 0x104=USB,MAGICWAKE
   id=515  level=255 0x100=MAGICWAKE creat=12/3/21, 11:53 AM  mod=12/3/21, 11:55 AM description=en0 owner=IOSkywalkNetworkBSDClient
   id=516  level=255 0x100=MAGICWAKE creat=12/3/21, 11:57 AM  mod=12/3/21, 4:34 PM description=llw0 owner=IOSkywalkNetworkBSDClient
   id=694  level=255 0x4=USB creat=12/3/21, 6:47 PM description=com.apple.usb.externaldevice.02100000 owner=IOUSBHostDevice
   id=696  level=255 0x4=USB creat=12/3/21, 6:51 PM description=com.apple.usb.externaldevice.02130000 owner=4-Port USB 2.1 Hub
   id=698  level=255 0x4=USB creat=12/3/21, 6:39 PM description=com.apple.usb.externaldevice.02131000 owner=Ducky Keyboard

**Comment by anirudhsama** [comment]
I started noticing this issue when I upgraded to the new M1 MacBook Pros. I use Raycast and if run the sleep command from the inbuilt keyboard, it says asleep. But running the same command from the bluetooth keyboard immediately wakes up the Mac. If I quit Karabiner, then the Mac sleeps properly.

**Comment by isametry** [comment]
Hello, Karabiner also prevents me from putting my display to sleep. Specifically, if my screen is locked and I want to press `esc` to turn it off, it always turns itself back on after less than a second.

Interestingly, this only affects the `esc` key on the built-in keyboard. Clicking `esc` on an external keyboard works fine, just as clicking the on-screen  "cancel" button with a mouse or trackpad.

I'm positive it's Karabiner causing this. Creating a new profile doesn't help.

MacBook Air (M1 2020)
Built-in display – no external monitors
Karabiner 14.3.0
Monterey 12.1

**Comment by net** [comment]
From my testing, Karabiner will cancel sleep if sleep is triggered by pressing a key on a Karabiner-controlled keyboard.

For example, when triggering sleep through Raycast or Alfred, pressing enter to activate the relevant sleep command will cause macOS to sleep then immediately wake; clicking (in Raycast you can click on Open Command) to activate the sleep command will cause macOS to sleep and stay asleep.

The same rules explain @isametry's observations: exit the lock screen by pressing escape and macOS will sleep then wake, exit by clicking Cancel and macOS will stay asleep.

The only exception to this I can find is that highlighting Sleep from the Apple menu then pressing enter will cause macOS to sleep and _stay asleep_, and I believe I know why.

I'm guessing there's an event that occurs shortly after the first key down event triggers the system's sleep that causes the system to wake.

**If I press and release the triggering key fast enough**—and I mean _truly_ fast, so that there is only a few ms between the key down and key up events—when triggering a sleep, then macOS will stay asleep.

We can test this in the command line. Pressing enter to run `pmset sleepnow` will exhibit the unwanted sleep-wake behavior. `sleep 0.1; pmset sleepnow` will exhibit the behavior if I don't release enter within 0.1 seconds, `sleep 0.2; pmset sleepnow` will exhibit the behavior if I don't release enter within 0.2 seconds, and so on.

This is why triggering sleep by pressing enter in the Apple menu doesn't exhibit the behavior; an animation first plays in the menu, _then_ the underlying action is triggered, by which time the key has been released. If we hold down the enter key when using the Apple menu longer than usual, we see the familiar unwanted sleep-wake behavior.

To recap, the system will wake when:

1. key down happens
2. underlying sleep command is triggered
3. ...?
4. key up happens

The system will not wake when:

1. key down happens
2. ...?
3. key up happens
4. underlying sleep command is triggered

I don't know what causes this behavior to exhibit with some input devices and not others. For me, it occurs with any keyboard controlled by Karabiner—in my case an M1 14-inch MacBook Pro internal keyboard and a Magic Keyboard. Removing a device from Karabiner's control (Devices > Basic Configuration > Modify events from this device) will prevent this behavior.

@tekezo I wish I could hunt further, but unfortunately this is the limit of my familiarity with the systems involved. I'm hoping this will be enough for you to take another look at this issue.

**Comment by jason5ng32** [comment]
+1

**Comment by sn0wyfall** [comment]
Can we please get a solution to this fucking issue? It's been plaguing Karabiner for a long time now. I can't believe that it's still happening in _Monterey_. I don't give a shit about new features. Just fix the bugs first.

**Comment by FelixLisczyk** [comment]
If it helps, my current workaround is to lock the screen with Ctrl + Cmd + Q or the Touch ID button. macOS goes to sleep automatically after a few seconds and doesn't wake up again on my device. I'm using a 14" MacBook Pro.

**Comment by isametry** [comment]
@sn0wyfall While I agree with you and share your frustration, making rude comments here doesn’t help anything.

This is a free, open-source tool – you aren’t entitled to using it or having it work correctly. The contributors don’t owe you anything.

If you know how to fix the issue, feel free to do so and submit a pull request.

**Comment by sn0wyfall** [comment]
> @sn0wyfall While I agree with you and share your frustration, making rude comments here doesn’t help anything.
> 
> This is a free, open-source tool – you aren’t entitled to using it or having it work correctly. The contributors don’t owe you anything.
> 
> If you know how to fix the issue, feel free to do so and submit a pull request.

I know it's a free tool. If Karabiner Elements was abondonware then I wouldn't care. But it gets updates, so to me it's inexcusable that this type of an issue can continue on for over a year. How about not releasing new features and just fixing a small bug?

**Comment by MSch** [comment]
@sn0wyfall You make it out like the creator of Karabiner Elements owes you anything at all - they do not. Instead of complaining please go ahead and investigate and fix the problem. When you're done, make sure to me on the PR so I can let you know if you fixed my problem, or if you need to go back and do another round of investigation and bug-fixing. Thank you!

**Comment by rublev** [comment]
> > @sn0wyfall While I agree with you and share your frustration, making rude comments here doesn’t help anything.
> > This is a free, open-source tool – you aren’t entitled to using it or having it work correctly. The contributors don’t owe you anything.
> > If you know how to fix the issue, feel free to do so and submit a pull request.
> 
> I know it's a free tool. If Karabiner Elements was abondonware then I wouldn't care. But it gets updates, so to me it's inexcusable that this type of an issue can continue on for over a year. How about not releasing new features and just fixing a small bug?

Bro go fork it and make the change lol, are you paying for it? No?

**Comment by jottr** [comment]
> But it gets updates, so to me it's inexcusable that this type of an issue can continue on for over a year. How about not releasing new features and just fixing a small bug?

Since it's a small issue, you should've been able to provide a patch in days yourself.
You should've had ample time to read the code, investigate the bug and provide a fix by now.
When can we expect a solution from you @sn0wyfall?

**Comment by jukrb0x** [comment]
> M1 Macbook Pro 16-inch macOS Monterey

Creating a new profile, then the problem went off **at the first time**, then I modified the binding, and the problem occurred again and since then the problem is persistent.

some alike issue, but nor sleeping, I meet the problem when **locking the screen** since when I went to bed and I caught the screen **doesn't go off for a long time**. I am pretty sure this is caused by Karabiner.

When I am on the locking screen, if I press `esc` on a kara-modified keyboard _(K2 in the below screenshot)_, the screen went black but **will awake immediately.**

the built-in keyboard is not controlled by Karabiner so I can lock the screen well when pressing `esc`:

![image](https://user-images.githubusercontent.com/15688641/180494082-60f773b6-ab8d-4e2f-92c6-27a93bbf1567.png)

**Comment by yusuketomoto** [comment]
My workaround was to create a simple script command which delays decent moment and then sleep. I'm using Raycast but I think that this way will work on Alfred as well. Disabling default Sleep command for Raycast would also help so that you can do it in the same way as before.

```
#!/usr/bin/osascript

# Required parameters:
# @raycast.schemaVersion 1
# @raycast.title Sleep
# @raycast.mode silent
# @raycast.packageName System

# Optional parameters:
# @raycast.icon :zzz:

# Documentation:
# @raycast.description Put your Mac to sleep.

delay 0.1
tell application "Finder" to sleep
```

**Comment by oakif** [comment]
In my opinion, this is probably the most significant bug in Karabiner. My MacBook has not had its display turned off in a few years. I wish this issue was addressed.

**Comment by cztomsik** [comment]
Ok, so maybe small update from me: it might be because of Bluetooth mouse because I don't have this issue anymore, and I didn't have it on the previous machine either. Maybe can you try to disable Bluetooth and test it again?

**Comment by oakif** [comment]
> Ok, so maybe small update from me: it might be because of Bluetooth mouse because I don't have this issue anymore, and I didn't have it on the previous machine either. Maybe can you try to disable Bluetooth and test it again?

Thanks for your trying this out! I disabled my Bluetooth but unfortunately, it didn't solve the issue. When I quit Karabiner the issue is fixed. I'm on the bus right now and whenever I lock my screen it just stays on, even when I'm on battery. Probably the most unfortunate bug I've ever run into during my years of using macOS

**Comment by aspacca** [comment]
I've installed Karabiner a few days ago: since then when I lock the screen and press `ESC` to let it suspend, it goes back to the login screen as soon as I release they `ESC` key.
If I quit Karabiner everything is working properly.

@tekezo , since the issue is open since Nov 2020, if you have at least an hint on how to fix this or what's causing this and can point me in the right direction, I will try to open a PR :)

thanks

**Comment by oakif** [comment]
It would be great if someone could have a look! I am a little scared of burn-in issues, particularly with my MacBook Pro 2021 with the mini-LED display, not to mention all the battery issues from having the screen be on all the time!

**Comment by oakif** [comment]
It would be great if someone could have a look! I am a little scared of burn-in issues, particularly with my MacBook Pro 2021 with the mini-LED display, not to mention all the battery issues from having the screen be on all the time!

**Comment by aspacca** [comment]
@oakif see my efforts on https://github.com/pqrs-org/Karabiner-Elements/pull/3366


> am a little scared of burn-in issues, particularly with my MacBook Pro 2021 with the mini-LED display, not to mention all the battery issues from having the screen be on all the time!

I have to habit to always lock the screen as soon as I get away from the macbook pro, at least for this there's a workaround: click with the pointer device on "cancel" icon (not sure about the english exact label) in the middle bottom of the screen, on the left from "change user" icon

**Comment by oakif** [comment]
> I have the habit to always lock the screen as soon as I get away from the macbook pro, at least for this there's a workaround: click with the pointer device on "cancel" icon (not sure about the english exact label) in the middle bottom of the screen, on the left from "change user" icon

Wow, thank you for this workaround. I used to use the "Escape" key to turn the screen off, thinking that it had the same effect as clicking that "cancel" button. However, the escape key causes the screen to turn back on again, while clicking the cancel button keeps the screen off. Cheers, and thank you for looking into the issue!

**Comment by YihangFu** [comment]
Does anyone know how to fix it now?

**Comment by submarines-and** [comment]
You can run @yusuketomoto scripts directly in karabiner using `to_delayed_action`. This works for me (change key command to whatever you want of course).

```
 {
    "description": "Sleep (cmd+opt+backspace)",
    "manipulators": [
        {
            "from": {
                "key_code": "delete_or_backspace",
                "modifiers": {
                    "mandatory": [
                        "command",
                        "option"
                    ]
                }
            },
            "to_delayed_action": {
                "to_if_invoked": [
                    {
                        "shell_command": "osascript -e 'tell application \"Finder\" to sleep'"
                    }
                ]
            },
            "type": "basic"
        }
    ]
},
```

**Comment by dew0** [comment]
> You can run @yusuketomoto scripts directly in karabiner using `to_delayed_action`. This works for me (change key command to whatever you want of course).
> 

It worked! Nice fix, thanks so much, this was driving me crazy!

**Comment by krizajb** [comment]
Works for me, fingers crossed it persists. Will keep you updated in the case it stops working.

**Comment by NightMachinery** [comment]
> @oakif see my efforts on #3366
> 
> > am a little scared of burn-in issues, particularly with my MacBook Pro 2021 with the mini-LED display, not to mention all the battery issues from having the screen be on all the time!
> 
> I have the habit to always lock the screen as soon as I get away from the macbook pro, at least for this there's a workaround: click with the pointer device on "cancel" icon (not sure about the english exact label) in the middle bottom of the screen, on the left from "change user" icon

Is this cancel icon available in macOS Sonoma? I cannot find anything. `ESC` doesn't work, as others have said.

**Comment by NightMachinery** [comment]
> @sn0wyfall While I agree with you and share your frustration, making rude comments here doesn’t help anything.
> 
> This is a free, open-source tool – you aren’t entitled to using it or having it work correctly. The contributors don’t owe you anything.
> 
> If you know how to fix the issue, feel free to do so and submit a pull request.

Can't we open an issue on Apple's part? Keyboard customization should ideally be an OS feature. It should even be legally required for accessibility alone. 

I myself only use Karabiner to map the FN key to press all the other modifiers together. I map all other hotkeys using Hammerspoon, which is higher-level and doesn't cause weird issues.

---

I have found a workaround to putting the display to sleep. Use `Hot Corners Shortcuts` in the OS settings. Since it uses only the mouse, the display stays off.

**Comment by cztomsik** [comment]
> I have found a workaround to putting the display to sleep. Use Hot Corners Shortcuts in the OS settings. Since it uses only the mouse, the display stays off.

Oh, that explains a lot :) I've been doing this for years, without any problem. Sorry if I've misguided someone here, this totally works.

**Comment by nejmeijer** [comment]
Thanks for reporting this, @Deftknot and others. The maintainer is not ignoring this issue: there are simply more than a thousand other open issues that also need attention. Others forms of help to solve this sooner in the form of PRs for tests or even failed attempts to fix this etc. would be highly appreciated.

**Comment by aspacca** [comment]
this was solved with https://github.com/pqrs-org/Karabiner-Elements/pull/3369, wasn't it?

**Comment by nejmeijer** [comment]
The last commit on that issue seems to indicate a better solution would still be desirable, but since you have more insight in this bug I think you are free to close this issue as you see fit @aspacca

**Comment by tekezo** [comment]
This issue should be resolved by #3369, so I close it.
As a side effect of the fix, the device needs to be re-grab upon wakeup because it is released right before sleep. However, I believe this is currently the best approach.
